ApiDiff: Add AssemblySymbolLoader method, a package dependency from CodeAnalysis, and minor fixes

carlossanlop opened the pull request:
This PR is part of the work needed to create an ApiDiff tool that reuses some of the code from Microsoft.DotNet.GenAPI. The idea is to make the larger PR smaller and make it easier to review: https://github.com/dotnet/sdk/pull/45389

The purpose of this change is to:

- Add the dependency to Microsoft.CodeAnalysis.Workspaces.Common, as it will be used in the upcoming change. The files adding this dependency are often showing merge conflicts due to the packages above and below changing often, making it very difficult to keep the PRs up to date.
- Add an API that will be later be used to refactor some of the other methods in the same class.
- Some minor fixes of comments and typos.

carlossanlop committed:
Add Microsoft.CodeAnalysis.Workspaces.Common package dependency.

carlossanlop committed:
Minor fixes.

carlossanlop committed:
Add IAssemblySymbol.LoadAssembliesAsDictionary.

ViktorHofer commented on src/SourceBuild/content/test/Microsoft.DotNet.UnifiedBuild.Tests/NugetPackageDownloadHelpers.cs:
This is a file outside of the Compatibility code base which we don't directly own. I would revert these changes.

carlossanlop replied:
I made this change since we were talking the other day about being more mindful of consistency across the repo. It's also a minor change that does not modify behavior.

ViktorHofer reviewed (commented):
Can you please describe why the new AssemblySymbolLoader API is necessary?

carlossanlop committed:
revert change to unrelated file

carlossanlop committed:
Add AssemblySymbolLoader factory method that consumes LoadAssembliesAsDictionary.

carlossanlop commented:
> Can you please describe why the new AssemblySymbolLoader API is necessary?

Sure: In the original PR I added a factory method to create an AssemblySymbolLoader instance. It uses this `LoadAssembliesAsDictionary` method. I added this factory method to the PR.

But I also noticed that I have simplified the ApiDiff tool enough to not need this method anywhere else. So I'll make it private, as I still need it for the factory method.

carlossanlop committed:
Make LoadAssembliesAsDictionary

carlossanlop committed:
Revert changes to unrelated file (for real now)

ViktorHofer commented:
I was talking about this new public static method: `public static (AssemblySymbolLoader, Dictionary<string, IAssemblySymbol>) CreateFromFiles(ILog logger, string[] assembliesPaths, string[]? assemblyReferencesPaths, bool respectInternals = false)`.

I would like to understand why AsmDiff needs a new entry point to `AssemblySymbolLoader` given that APICompat and GenAPI don't require it.

carlossanlop commented:
I'll split this. Some things can be sent separately alongside a bigger change.
